### DISCUSSION TITLE
Import library on both server and client

### DISCUSSION THREAD

**Discussion by simplecommerce** [vikejs/vike#628] [Discussion] [answered]
Hi,

I am new to this and I am trying to test the migration of an existing app to this.

I have a custom library that I have to import and initialize on the server and client before any pages or any other code is executed.

I am not sure where I should put this.

I thought I could simply place them above in my _default.page.server.jsx and _default.page.client.jsx  but I am not sure thats where it belongs.

Any help is appreciated.
One thing I haven't seen in the documentation or I haven't found, was the order of the loading in the file structure.

**Comment by brillout** [comment, marked as answer]
Usually in your server entry (e.g. your Express.js code), but `_default.page.server.js` can be fine as well; I'd do whatever works best.

**Reply by simplecommerce** [reply]
I need both my server and client to import the library first though.
I tried in the `_default.page.server.js` and `_default.page.client.js` but they seem to be imported after the page is imported.

Thanks for your reply!

**Reply by brillout** [reply]
If you import the library's server-side at your server entry, then you can it to be the very first thing that is being loaded in Node.js.

For the library's client-side.

Either you can inject the library's client-side directly in the HTML and guarantee it's the first thing being loaded, or you add it in `_default.page.client.js` but then, indeed, it won't be the first line of JS of code loaded in the browser. Feel free to let me know if that's a blocker.

**Reply by simplecommerce** [reply]
Hmm, yeah it's because I am using React for the front-end and we have a custom library that we load our generic components, code, etc.
And it order to make it available, we import them on the client/server startup sequence to initialize our library and global stuff.
And then in our code we can access them by calling our library.
With MeteorJS I was able to do this by simply adding my file in the client/main.js and server/main.js startup files.
But when I tried to do it with vite-plugin-ssr, I wasn't able to and wasn't sure how to do it properly to keep the same logic.
I will try to play around with it more and see if I can come up with something that works.

Thanks!

**Reply by brillout** [reply]
AFAICT my previous reply should work out with what you want to achieve.

> I will try to play around with it more and see if I can come up with something that works.

Let me know how that goes.

**Reply by simplecommerce** [reply]
I know it has been a long time but I figured i'd give an update.
I tried recently to do this with my project in typescript and I couldn't get it working by putting the server side import of my library in the `/server.index.ts` because it didn't like my import which was `import "./lib" and inside I had a `indes.ts` it wouldnt detect it and gave me a module not found error.

So I tried in the `_default.page.server.js` and that didn't work also because it wouldn't import it before importing my pages that depend on this library.

So my temporary fix was to do the import inside my library that is imported by all my pages. But this doesn't seem efficient either.

I tried to do something like

```
if (import.meta.env.SSR) await import("./lib/server.ts")
else await import("./lib/client.ts");
```

This only seemed to work on server side but on client side it wouldn't work at all.
On the client side I think I have to do `import "./lib/client.ts"` at the very top for it to work properly, but I wasn't able to do this at the moment.

So I am considering changing the way I import and include my library in my current project that I am testing to get it working under this library.

I didn't know how to inject my library code on the client side apart from the normal import I do.

**Reply by brillout** [reply]
What is `lib` doing?

**Reply by simplecommerce** [reply]
It's just a custom library that we built to load components, functions in a global object that we import/use in our pages/components in order to not have multiple imports to make it simpler if we need to change the internal librairies.

For example:

// in our definitions
```
import Lib from "/imports/lib";

const SomeComponent = () => <div />;

Lib.addComponent("SomeComponent", SomeComponent);
```

// in our page
```
import Lib from "/imports/lib";

const { SomeComponent } = Lib.getComponents();

const Page = () => <SomeComponent />
```

Sorry if I am unclear.

**Reply by brillout** [reply]
> One thing I haven't seen in the documentation or I haven't found, was the order of the loading in the file structure.

I think it's best to not assume any import order. In general, with JavaScript, I don't find myself needing a deterministic import order.

So I'm not sure why you need this, and I'm inclined to think that there should be a solution to your problem in user land that doesn't assume anything about the import order.

That said, if you do believe there is something vite-plugin-ssr should be doing differently, feel free let me know.

There is https://github.com/brillout/vite-plugin-ssr/issues/962 but it's about execution order not import order.

Btw: would your company be up for [sponsoring](https://github.com/sponsors/brillout)?

**Reply by simplecommerce** [reply]
Hmm the hook might be useful in this case if I can import required libs before any other page is rendered.
But yeah, I think I might have to change the way my project imports its library in order to get it working.
I haven't had time to try to change my current project and adapt it to your library yet, I just started to do some tests now so I haven't had a chance to get it fully working yet to see if its viable. 

I am in the beginning stages to try and see if its viable to switch or not.

**Reply by simplecommerce** [reply]
Ok so I made some progress, I was able to get it sort of working on the server side.
I had to change my imports and using only import / exports in order to get it to work, so dropping our library functions.

But for some reasons while on the server side it worked, on the client side I kept getting errors like "cannot use Component before its initialization" i forgot the exact error, but I switched my import statement from `import { components } from "/imports"` to const { components } = await import("/imports")` and it seemed to rermove the error for now.

But for some reason, my render function on the client never gets called event if `export clientRouting = true` is uncommented.
I do see the console.log I put outside of the render function so I know the file is being called.
I need to do more debugging, I probably have an error somewhere that I am not seeing in the console.

**Reply by brillout** [reply]
👍 Yea it all seems to be user land issues.

**Reply by simplecommerce** [reply]
Ok so I got it somewhat working, at least I can see my login page and it seems to work on the client side. 
It was quite difficult to figure out the issues because I had so many things to change and in the end I didn't need to change my import to `await import`, it was an underlying code that I had to fix but there wasn't any error messages that hinted in that direction so I could properly go and fix them but in the end I was able to.
I'll do some more tests and see how far I can take it but so far it seems to work!
